synapse_patch_push_rules/__init__.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2022 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from typing import Any, Dict, List, Union
+
+import attr
+from synapse.module_api import ModuleApi
+from synapse.module_api.errors import ConfigError
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s(auto_attribs=True, frozen=True)
+class PushRule:
+    kind: str
+    actions: Dict[str, List[Union[str, Dict[str, str]]]]
+
+
+class PushRulesPatcherConfig:
+    def __init__(self, config: Dict[str, Any]) -> None:
+        self.rules: Dict[str, PushRule] = {}
+        for rule_id, rule in config["rules"].items():
+            self.rules[rule_id] = PushRule(**rule)
+
+
+class PushRulesPatcher:
+    def __init__(self, config: PushRulesPatcherConfig, api: ModuleApi):
+        # Keep a reference to the config and Module API
+        self._api = api
+        self._config = config
+
+        # Check that the actions are valid.
+        for rule_id, rule in self._config.rules.items():
+            if not rule_id.startswith("."):
+                raise ConfigError(
+                    "Only the rules predefined in the Matrix specification are supported"
+                    " by this module. See https://spec.matrix.org/latest/client-server-api/#predefined-rules"
+                    " for a complete list."
+                )
+
+            if self._api.check_push_rule_actions(rule.actions) is False:
```

Comment:
style nit: 
```suggestion
            if not self._api.check_push_rule_actions(rule.actions):
```
would be more idiomatic since `check_push_rule_actions` only returns True or False

Ignore this comment if we decide to make `check_push_rule_actions` return an `InvalidRuleException`

Reply:
Changed to a try/catch clause now that https://github.com/matrix-org/synapse/pull/12406 makes `check_push_rule_actions` raise if invalid.